File: java/src/main/java/problems/Problem009LinkedListCycle.java
package problems;


public class Problem009LinkedListCycle {
    static class Node{int v; Node next; Node(int v){this.v=v;}}
    public static boolean hasCycle(Node h){
        Node s=h,f=h; while(f!=null&&f.next!=null){ s=s.next; f=f.next.next; if(s==f) return true; }
        return false;
    }
}

File: scala/src/main/scala/problems/Problem009LinkedListCycle.scala

package problems

object Problem009LinkedListCycle {
  final class Node(var v:Int, var next: Node)
  object Node { def apply(v:Int): Node = new Node(v, null) }

  def hasCycle(h: Node): Boolean = {
    var slow = h; var fast = h
    while (fast!=null && fast.next!=null) {
      slow = slow.next; fast = fast.next.next
      if (slow eq fast) return true
    }
    false
  }
}

